smsdb.c

```diff
@@ -310,48 +387,362 @@ EXPORT_DEF int smsdb_put(const char *id, const char *addr, int ref, int parts, i
+	} else {
+		strcpy(dst, (const char*)sqlite3_column_text(get_payload_stmt, 1));
+		res = sqlite3_column_bytes(get_payload_stmt, 0);
+		res = res > 4095 ? 4095 : res;
+		memcpy(payload, sqlite3_column_blob(get_payload_stmt, 0), res);
+		payload[res] = '\0';
+	}
+	sqlite3_reset(get_payload_stmt);
+
+	if (res != -1 && smsdb_outgoing_clear_nolock(uid) < 0) {
+		res = -1;
+	}
+
+	db_sync();
+	ast_mutex_unlock(&dblock);
+
+	return res;
+}
+EXPORT_DEF ssize_t smsdb_outgoing_part_put(int uid, int refid, char *dst, char *payload)
+{
+	int res = 0;
+	char fullkey[MAX_DB_FIELD];
+	size_t fullkey_len;
+	int srr = 0, cnt, cur;
+
+	ast_mutex_lock(&dblock);
+
+	if (sqlite3_bind_int(get_outgoingmsg_stmt, 1, uid) != SQLITE_OK) {
+		ast_log(LOG_WARNING, "Couldn't bind UID to stmt: %s\n", sqlite3_errmsg(smsdb));
+		res = -1;
+	} else if (sqlite3_step(get_outgoingmsg_stmt) != SQLITE_ROW) {
+		res = -2;
+	} else {
+		const char *dev = (const char*)sqlite3_column_text(get_outgoingmsg_stmt, 0);
+		const char *dst = (const char*)sqlite3_column_text(get_outgoingmsg_stmt, 1);
+		srr = sqlite3_column_int(get_outgoingmsg_stmt, 2);
+		if (strlen(dev) + strlen(dst) + 3 + 2 >= sizeof(fullkey)) {
+			ast_log(LOG_WARNING, "Key length must be less than %zu bytes\n", sizeof(fullkey));
+			res = -1;
+		}
+
+		fullkey_len = snprintf(fullkey, sizeof(fullkey), "%s/%s/%d", dev, dst, refid);
+	}
+	sqlite3_reset(get_outgoingmsg_stmt);
+
+	if (res != -1) {
+		if (sqlite3_bind_text(put_outgoingpart_stmt, 1, fullkey, fullkey_len, SQLITE_STATIC) != SQLITE_OK) {
+			ast_log(LOG_WARNING, "Couldn't bind key to stmt: %s\n", sqlite3_errmsg(smsdb));
+			res = -1;
+		} else if (sqlite3_bind_int(put_outgoingpart_stmt, 2, uid) != SQLITE_OK) {
+			ast_log(LOG_WARNING, "Couldn't bind UID to stmt: %s\n", sqlite3_errmsg(smsdb));
+			res = -1;
+		} else if (sqlite3_step(put_outgoingpart_stmt) != SQLITE_DONE) {
+			res = -1;
+		}
+		sqlite3_reset(put_outgoingpart_stmt);
+	}
+
+	if (srr) {
+		res = -2;
+	}
+
+	// if no status report is requested, just count successfully inserted parts and return payload if the counter reached the number of parts
+	if (res >= 0) {
+		if (sqlite3_bind_int(cnt_all_outgoingpart_stmt, 1, uid) != SQLITE_OK) {
+			ast_log(LOG_WARNING, "Couldn't bind key to stmt: %s\n", sqlite3_errmsg(smsdb));
+			res = -1;
+		} else if (sqlite3_step(cnt_all_outgoingpart_stmt) != SQLITE_ROW) {
+			res = -1;
+		} else {
+			cur = sqlite3_column_int(cnt_all_outgoingpart_stmt, 0);
+			cnt = sqlite3_column_int(cnt_all_outgoingpart_stmt, 1);
+		}
+		sqlite3_reset(cnt_all_outgoingpart_stmt);
+	}
+
+	if (res >= 0 && cur != cnt) {
+		res = -2;
+	}
+
+	// get payload
+	if (res >= 0) {
+		if (sqlite3_bind_int(get_payload_stmt, 1, uid) != SQLITE_OK) {
+			ast_log(LOG_WARNING, "Couldn't bind key to stmt: %s\n", sqlite3_errmsg(smsdb));
+			res = -1;
+		} else if (sqlite3_step(get_payload_stmt) != SQLITE_ROW) {
+			res = -1;
+		} else {
+			strcpy(dst, (const char*)sqlite3_column_text(get_payload_stmt, 1));
```

Comment:
Like mentioned in at_response.c, you're writing to dst without knowing if it fits. (Also don't forget about the NULs.)